[Automatic Migrations] Support for QRadar Mitre Mappings via Rule Enhancements

logeekal opened the pull request:
## Summary

- Fixes https://github.com/elastic/security-team/issues/14587

This PR Adds a new `enhance` endpoint which let's user add data the existing rules. There may be cases where all rules data is not contained in single file but is divided in multiple files. Enhance endpoint helps to combine that data. 

Current use case to add this data is to add MITRE mappings to existing rules since those mappings could be contained in a separate file.

Below Are details of the endpoints


### POST /internal/siem_migrations/rules/{migration_id}/rules/enhance

*Request Body*

```js
{
   "vendor" : "qradar",
    // enhancement type
   "type" : "mitre",
   "data" :  {"mitre" : "json"}
}

```

*Response*
```js

// 200 OK
{
   "updated" : "true"
}
```






## Desk Testing Guide

Before starting anything, please download given files.

- [sample_rules.xml](https://github.com/user-attachments/files/23630853/sample_rules.xml)
- [sample_rule_mitre_mapping.json](https://github.com/user-attachments/files/23630854/sample_rule_mitre_mapping.json)




1. Create a Rule Migration 

```sh
curl --request PUT \
  --url http://localhost:5601/internal/siem_migrations/rules \
  --header 'authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'content-type: application/json' \
  --header 'elastic-api-version: 1' \
  --header 'kbn-xsrf: true' \
  --header 'x-elastic-internal-origin: security-solution' \
  --data '{
  "name": "My Migration with SAMPLE Qradar Rule"
}'
```

2. In Step 1, you will receive a migration_id. Let's called it `<current_migration_id>`. Upload the XML File using below commands. Replace path to the file and migration_id appropriately.

```bash
xml_content=$(cat /path/to/sample_rules.xml | tr -d '\n' | sed 's/"/\\"/g')
json_payload=$(printf '{"xml": "%s"}' "$(echo "$xml_content" )")
curl --request POST --url http://localhost:5601/internal/siem_migrations/rules/<current_migration_id>/qradar/rules --header 'authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' --header 'content-type: application/json' --header 'elastic-api-version: 1' --header 'kbn-xsrf: true ' --header 'x-elastic-internal-origin: security-solution' --data "$json_payload"
```

You will recieve below response confirming that importing QRadar rule has been successful.

3. Upload The Mitre Mapping file with below format. 

```sh
curl --request POST \
  --url http://localhost:5601/internal/siem_migrations/rules/<current_migration_id>/rules/enhance \
  --header 'authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'content-type: application/json' \
  --header 'elastic-api-version: 1' \
  --header 'kbn-xsrf: true' \
  --header 'x-elastic-internal-origin: security-solution' \
  --data '{
  "vendor": "qradar",
  "type": "mitre",
  "data": {
    "A sample QRadar Rule": {
      "id": "b53c0e32-f942-4f01-bff3-e64b60e176dd",
      "has_ibm_default": false,
      "last_updated": 1717423281512,
      "mapping": {
        "TA0001": {
          "confidence": "medium",
          "user_override": false,
          "enabled": true,
          "ibm_default": false,
          "name": "Initial Access",
          "techniques": {
            "T1190": {
              "confidence": "medium",
              "enabled": true,
              "id": "T1190"
            }
          }
        },
        "TA0008": {
          "confidence": "medium",
          "user_override": false,
          "enabled": true,
          "ibm_default": false,
          "name": "Lateral Movement",
          "techniques": {
            "T1210": {
              "confidence": "medium",
              "enabled": true,
              "id": "T1210"
            }
          }
        }
      },
      "min-mitre-version": 8
    }
  }
}'
```

4. Once MITRE Mappings have been imported. Start the Migration from UI and check that the resulting Rule will have the same mitre mappings as shown in below screenshot.

<img width="1058" height="548" alt="image" src="https://github.com/user-attachments/assets/df43e9da-aa9b-4fb5-80b1-2d24e0f4503d" />



### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [ ] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

### Identify risks

Does this PR introduce any risks? For example, consider risks like hard to test bugs, performance regression, potential of data loss.

Describe the risk, its severity, and mitigation for each identified risk. Invite stakeholders and evaluate how to proceed before merging.

- [ ] [See some risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx)
- [ ] ...

logeekal committed:
Mitre mappings : initial implementation

logeekal committed:
fix: vendor processing logic + integration tests

logeekal committed:
fix: types

logeekal committed:
fix: more types

logeekal committed:
Fix: Qradar transforms and types

logeekal committed:
fix: add tactic without technique

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/363140)
* Commit: e0db4eb59830feb3eabd17ecd3673e2ce354b103

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/363140#019aa113-a972-4f24-940a-11f58d3d78ec)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/363140","number":363140,"commit":"e0db4eb59830feb3eabd17ecd3673e2ce354b103"}],"number":363140}
buildkite-pr-comment-->

logeekal committed:
fix: types + tests

logeekal committed:
fix: types

logeekal committed:
Merge branch 'main' into feat/qradar_mitre_mappings

kqualters-elastic commented on x-pack/solutions/security/plugins/security_solution/package.json:
```diff
@@ -35,5 +35,8 @@
     "openapi:bundle:endpoint-management": "node scripts/openapi/bundle_endpoint_management"
+  },
+  "devDependencies": {
+    "baseline-browser-mapping": "^2.8.29"
```
what's this for?

logeekal committed:
fix: address PR review comments

- Remove unused baseline-browser-mapping devDependency
- Enable all siem_migrations integration tests
- Add qradarRulesMigration feature flag to serverless config

angorayc commented:
> Upload The Mitre Mapping file with below format.
> ```
> curl --request POST \
>   --url http://localhost:5601/internal/siem_migrations/rules/<current_migration_id>/rules/enhance \
>   --header 'authorization: Basic {AUTH}' \
>   --header 'content-type: application/json' \
>   --header 'elastic-api-version: 1' \
>   --header 'kbn-xsrf: true' \
>   --header 'x-elastic-internal-origin: security-solution' \
>   --data '{
>   "vendor": "qradar",
>   "enhancement_type": "mitre",
>   "data": {
>     "A sample QRadar Rule": {
>       "id": "b53c0e32-f942-4f01-bff3-e64b60e176dd",
>       "has_ibm_default": false,
>       "last_updated": 1717423281512,
>       "mapping": {
>         "TA0001": {
>           "confidence": "medium",
>           "user_override": false,
>           "enabled": true,
>           "ibm_default": false,
>           "name": "Initial Access",
>           "techniques": {
>             "T1190": {
>               "confidence": "medium",
>               "enabled": true,
>               "id": "T1190"
>             }
>           }
>         },
>         "TA0008": {
>           "confidence": "medium",
>           "user_override": false,
>           "enabled": true,
>           "ibm_default": false,
>           "name": "Lateral Movement",
>           "techniques": {
>             "T1210": {
>               "confidence": "medium",
>               "enabled": true,
>               "id": "T1210"
>             }
>           }
>         }
>       },
>       "min-mitre-version": 8
>     }
>   }
> }'
> ```

"enhancement_type": "mitre", => Should be `type`: "mitre" to work properly

logeekal committed:
fix: rule installation

logeekal committed:
Merge branch 'main' into feat/qradar_mitre_mappings
